include/particles/particle_base.h

```diff
@@ -315,6 +348,107 @@ class ParticleBase {
       const std::vector<uint8_t>& buffer,
       std::vector<std::shared_ptr<mpm::Material<Tdim>>>& materials) = 0;
 
+  //! TwoPhase functions--------------------------------------------------------
+  //! Update porosity
+  //! \param[in] dt Analysis time step
+  virtual void update_porosity(double dt) {
+    throw std::runtime_error(
```

Comment:
Why do we have to do this? I would keep this pure virtual and throw error elsewhere?

Reply:
That means we need to populate the `particle.tcc` with all these functions. Are you okay with that?